File: train/322.txt.pas
Var
 da,a:array[0..30] of integer;
 str,str1:string;
 summ,sum,ch,max,i,s,data,code,abc,aa:integer;
Procedure dfs(k,n,tall:integer);
var
 i:integer;
begin
 if k>aa then
 begin
if n>summ then summ:=n;
exit;
 end;
 if da[k]<=tall then dfs(k+1,n+1,da[k]);
 dfs(k+1,n,tall);
end;

Begin
 readln(str);
 max:=1;
 a[1]:=maxint-1;
 a[0]:=maxint;
 abc:=maxint;
 while str<>'' do
 begin

s:=pos(',',str);
if s=0 then
begin
  val(str,data,code);
  str:='';
end else
begin
  str1:=copy(str,1,s-1);
  val(str1,data,code);
  delete(str,1,s);
end;
inc(aa);
da[aa]:=data;
ch:=0;
for i:=1 to max do
  if (data<=a[i]) and (a[i]<a[ch]) then
ch:=i;
if ch=0 then
begin
  inc(max);
  inc(a[max],data);
end;
if ch>0 then
begin
  a[ch]:=data;
  inc(sum);
end;
 end;

 dfs(1,0,maxint);
 writeln(summ,',',max-1);
End.

File: train/322.txt.cpp
#include <iostream>
using namespace std;
int n=0, c[102]={0};
bool f[2002][2002] = {0};
int sum = 0;
int main()
{
    cin>>n;
    for (int i=1; i<=n; i++)
    {
        cin>>c[i];
        sum += c[i];
    }
    f[0][0] = true;
    for (int i=1; i<=n; i++)
        for (int j=sum/2; j>=0; j--)
            for (int k=j; k>=0; k--)
            {
                if (f[j][k]) continue;
                if (j-c[i]>=0)
                {
                    if (j-c[i]>=k)
                    {
                        if (f[j-c[i]][k])
                            f[j][k] = true;
                    }
                    else
                    {
                        if (f[k][j-c[i]])
                            f[j][k] = true;
                    }
                }
                if (k-c[i]>=0)
                    if (f[j][k-c[i]])
                        f[j][k] = true;
            }
    for (int i=sum/2; i>=1; i--)
       if (f[i][i])
       {
           cout<<i;
           exit(0);
       }
    cout<<"Impossible";
    return 0;
}

